### PR TITLE
`EqualsAvoidsNull` should flip arguments for constants

### PR DESCRIPTION
**Pull request by punkratz312**
Hi,

I might have found an issue in my local project where around 90 violations flagged by PMD are not being replaced correctly. Some of these violations seem related to a list and method arguments.

Could you please take over and enable the necessary tests to confirm if this is indeed an issue?

Thanks in advance!  

@timtebeek 
@knutwannheden 
@ggerbaud

### CODE REVIEW

File: src/main/java/org/openrewrite/staticanalysis/EqualsAvoidsNullVisitor.java [hunk 1]
```diff
@@ -60,9 +60,15 @@ public J visitMethodInvocation(J.MethodInvocation method, P p) {
         if (m.getSelect() != null &&
             !(m.getSelect() instanceof J.Literal) &&
             !m.getArguments().isEmpty() &&
-            m.getArguments().get(0) instanceof J.Literal &&
+            (m.getArguments().get(0) instanceof J.Literal || m.getArguments().get(0) instanceof J.FieldAccess) &&
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/staticanalysis/EqualsAvoidsNullVisitor.java:63]: This is the relevant change here; in theory that FieldAccess might itself be nullable, whereas the old left hand side is not. That makes this a challenging change to make reliably without potentially making things worse as well.

Comment (timtebeek) [reply src/main/java/org/openrewrite/staticanalysis/EqualsAvoidsNullVisitor.java]: Added a check for the flags of the field type; with that we're good to merge.

**Review by timtebeek — APPROVED** [review]
Thanks for the help here @punkratz312 ! Nice to see these cases covered as well.

**Comment by punkratz312** [comment]
wow so quick 🥇  well done 🍀

**Comment by pankratz76** [comment]
this is wrong imho

**Comment by timtebeek** [comment]
Ah sorry to hear; in which way? What kind of changes would you propose we make now?

**Comment by pankratz76** [comment]
Sorry for causing this issue. Using the recipes in the codebase itself should have revealed this. Ideally, there should be no dirty code on `main`, assuming all GitHub Actions are applied beforehand. We could then easily implement a check to ensure that the recipes are idempotent. So, running them after merging into `main` should reveal no changes, as they must be applied beforehand ideally.

I experienced a similar situation where the Maven call required multiple steps to cover everything, or it had already flipped. However, if that were the case, it would have also resulted in an infinite loop locally, which seems unlikely. I should have experienced this right away, as there is no exit from that condition.

**Comment by pankratz76** [comment]
> Addressing references is tricky and not covered by other projects like PMD and Checkstyle. Implementing this breaks the idempotency, causing: #472
> 
> Changing a reference with another reference makes no sense to me, which is this feature kind of about. Its like fixing one leak by taking from another one thats not leaking ATM. We just flip for infinity as we can not truly determine which one is null, only by changing the arguments (a reference with another reference).

no PMD is making the assumption so this feature seems correct.

https://github.com/openrewrite/rewrite-static-analysis/pull/473#issuecomment-2720640459

**Pull request merged**